StyleGradientImages never cache their image

weinig opened the pull request:
#### f88290ea41bc0d62dcb4a581dfbcd787d4664796
<pre>
StyleGradientImages never cache their image
<a href="https://bugs.webkit.org/show_bug.cgi?id=286186">https://bugs.webkit.org/show_bug.cgi?id=286186</a>

Reviewed by Simon Fraser.

StyleGradientImage backing store caching has been broken since 287234@main.
The fact we were never caching was also masking an issue where we would fail
to draw anything if a cacheable StyleGradientImage was being used as the
input to another StyleImage, such as StyleFilterImage.

* Source/WebCore/style/values/images/StyleGradient.cpp:
     - Fix check for cacheability by updating `none_of` -&gt; `all_of`.
     - Remove unnecessary check for containing the client. Since the
       gradient is cacheable and cannot change we don&apos;t care about
       the client here like we do for StyleCanvasImage where client
       must be notified on changes. Removing this check allows
       gradients used by other StyleImages (like StyleFilterImage)
       to be successfully cached.

* Source/WebCore/style/values/color/StyleColor.h:
    - Export some functions to make unit test work.

* Tools/TestWebKitAPI/CMakeLists.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/StyleGradient.cpp: Added.
    - Add unit tests for both cacheable and uncacheable states.

Canonical link: <a href="https://commits.webkit.org/289123@main">https://commits.webkit.org/289123@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/964ccc33bfcf10e5abdf6725adcdc5f98cfbb909

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/85387 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/5117 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39813 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90510 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36424 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87471 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/5223 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/13095 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66379 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24193 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88428 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3990 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77553 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46661 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3874 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31814 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35492 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74579 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32656 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/92034 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12730 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9319 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74992 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12958 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73390 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/74112 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18335 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18457 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16888 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/4771 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12688 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/18141 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12512 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15995 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/14268 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/style/values/color/StyleColor.h:
```diff
@@ -125,7 +125,7 @@ struct Color {
     Color& operator=(const Color&);
 
-    Color(Color&&);
+    WEBCORE_EXPORT Color(Color&&);
```
I always wonder if exporting functions makes them non-inlineable, and hurts perf.

weinig replied:
I don’t believe it does, but it sure is annoying.

weinig committed:
StyleGradientImages never cache their image
https://bugs.webkit.org/show_bug.cgi?id=286186

Reviewed by Simon Fraser.

StyleGradientImage backing store caching has been broken since 287234@main.
The fact we were never caching was also masking an issue where we would fail
to draw anything if a cacheable StyleGradientImage was being used as the
input to another StyleImage, such as StyleFilterImage.

* Source/WebCore/style/values/images/StyleGradient.cpp:
     - Fix check for cacheability by updating `none_of` -> `all_of`.
     - Remove unnecessary check for containing the client. Since the
       gradient is cacheable and cannot change we don't care about
       the client here like we do for StyleCanvasImage where client
       must be notified on changes. Removing this check allows
       gradients used by other StyleImages (like StyleFilterImage)
       to be successfully cached.

* Source/WebCore/style/values/color/StyleColor.h:
    - Export some functions to make unit test work.

* Tools/TestWebKitAPI/CMakeLists.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/StyleGradient.cpp: Added.
    - Add unit tests for both cacheable and uncacheable states.

Canonical link: https://commits.webkit.org/289123@main
